lib/puppet/provider/cisco_bridge_domain/cisco.rb

```diff
@@ -0,0 +1,147 @@
+begin
+  require 'puppet_x/cisco/autogen'
+rescue LoadError # seen on master, not on agent
+  # See longstanding Puppet issues #4248, #7316, #14073, #14149, etc. Ugh.
+  require File.expand_path(File.join(File.dirname(__FILE__), '..', '..', '..',
+                                     'puppet_x', 'cisco', 'autogen.rb'))
+end
+
+Puppet::Type.type(:cisco_bridge_domain).provide(:cisco) do
+  desc 'The new NXAPI provider.'
+
+  confine feature: :cisco_node_utils
+  defaultfor operatingsystem: :nexus
+
+  mk_resource_methods
+
+  BD_NON_BOOL_PROPS = [:bd_name]
+  BD_BOOL_PROPS = [:fabric_control, :shutdown]
+  BD_ALL_PROPS = BD_NON_BOOL_PROPS + BD_BOOL_PROPS
+
+  PuppetX::Cisco::AutoGen.mk_puppet_methods(:non_bool, self, '@bd',
+                                            BD_NON_BOOL_PROPS)
+  PuppetX::Cisco::AutoGen.mk_puppet_methods(:bool, self, '@bd',
+                                            BD_BOOL_PROPS)
+
+  def initialize(value={})
+    super(value)
+    @bd = Cisco::BridgeDomain.bds[@property_hash[:name]]
+    @property_flush = {}
+    debug 'Created provider instance of cisco_bridge_domain.'
+  end
+
+  def self.properties_get(bd_id, v)
+    debug "Checking instance, bd #{bd_id}"
+    current_state = {
+      bd:     bd_id,
+      name:   bd_id,
+      ensure: :present,
+    }
+
+    # Call node_utils getter for each property
+    BD_NON_BOOL_PROPS.each do |prop|
+      current_state[prop] = v.send(prop)
+    end
+    BD_BOOL_PROPS.each do |prop|
+      val = v.send(prop)
+      if val.nil?
+        current_state[prop] = nil
+      else
+        current_state[prop] = val ? :true : :false
+      end
+    end
+    new(current_state)
+  end # self.properties_get
+
+  def self.instances
+    bds = []
+    Cisco::BridgeDomain.bds.each do |bd_id, v|
+      bds << properties_get(bd_id, v)
+    end
+    bds
+  end
+
+  def self.prefetch(resources)
+    bds = instances
+
+    resources.keys.each do |id|
+      provider = bds.find { |bd| bd.instance_name == id }
+      resources[id].provider = provider unless provider.nil?
+    end
+  end # self.prefetch
+
+  def exists?
+    (@property_hash[:ensure] == :present)
+  end
+
+  def create
+    @property_flush[:ensure] = :present
+  end
+
+  def destroy
+    @property_flush[:ensure] = :absent
+  end
+
+  def instance_name
+    bd
+  end
+
+  def properties_set(new_bd=false)
+    BD_ALL_PROPS.each do |prop|
+      next unless @resource[prop]
+      send("#{prop}=", @resource[prop]) if new_bd
+      unless @property_flush[prop].nil?
+        @bd.send("#{prop}=", @property_flush[prop]) if
+          @bd.respond_to?("#{prop}=")
+      end
+    end
+  end
+
+  def flush
+    if @property_flush[:ensure] == :absent
+      @bd.destroy
+      @bd = nil
+    else
+      # Create/Update
+      if @bd.nil?
+        new_bd = true
+        @bd = Cisco::BridgeDomain.new(@resource[:bd])
+      end
+      properties_set(new_bd)
+    end
+    puts_config
+  end
+
+  def puts_config
```

Comment:
We've also found this `puts_config` to have dubious value. You can keep it but consider it optional. I haven't been using it in my own new providers.